prow/cmd/gangway/main.go

```diff
@@ -0,0 +1,225 @@
+	prowv1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	configflagutil "k8s.io/test-infra/prow/flagutil/config"
+	"k8s.io/test-infra/prow/gangway"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/metrics"
+
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	client         prowflagutil.KubernetesOptions
+	github         prowflagutil.GitHubOptions
+	port           int
+	cookiefilePath string
+
+	config configflagutil.ConfigOptions
+
+	dryRun                 bool
+	gracePeriod            time.Duration
+	instrumentationOptions prowflagutil.InstrumentationOptions
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.IntVar(&o.port, "port", 32000, "TCP port for gRPC.")
+	fs.BoolVar(&o.dryRun, "dry-run", true, "Dry run for testing. Uses API tokens but does not mutate.")
+	fs.DurationVar(&o.gracePeriod, "grace-period", 180*time.Second, "On shutdown, try to handle remaining events for the specified duration. ")
+	fs.StringVar(&o.cookiefilePath, "cookiefile", "", "Path to git http.cookiefile, leave empty for github or anonymous")
+	for _, group := range []flagutil.OptionGroup{&o.client, &o.github, &o.instrumentationOptions, &o.config} {
+		group.AddFlags(fs)
+	}
+
+	fs.Parse(args)
+
+	return o
+}
+
+func (o *options) validate() error {
+	var errs []error
+	for _, group := range []flagutil.OptionGroup{&o.client, &o.github, &o.instrumentationOptions, &o.config} {
+		if err := group.Validate(o.dryRun); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if o.port == o.instrumentationOptions.HealthPort {
+		errs = append(errs, fmt.Errorf("both the gRPC port and health port are using the same port number %d", o.port))
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+type kubeClient struct {
+	client prowv1.ProwJobInterface
+	dryRun bool
+}
+
+// Create creates a Prow Job CR in the Kubernetes cluster (Prow service cluster).
+func (c *kubeClient) Create(ctx context.Context, job *prowcr.ProwJob, o metav1.CreateOptions) (*prowcr.ProwJob, error) {
+	if c.dryRun {
+		return job, nil
+	}
+	return c.client.Create(ctx, job, o)
+}
+
+// interruptableServer is a wrapper type around the gRPC server, so that we can
+// pass it along to our own interrupts package.
+type interruptableServer struct {
+	grpcServer *grpc.Server
+	listener   net.Listener
+	port       int
+}
+
+// Shutdown shuts down the inner gRPC server as gracefully as possible, by first
+// invoking GracefulStop() on it. This gives the server time to try to handle
+// things gracefully internally. However if it takes too long (if the parent
+// context cancels us), we forcefully kill the server by calling Stop(). Stop()
+// interrupts GracefulStop() (see
+// https://pkg.go.dev/google.golang.org/grpc#Server.Stop).
+func (s *interruptableServer) Shutdown(ctx context.Context) error {
```

Comment:
@cjwagner This part is new vs https://github.com/kubernetes/test-infra/pull/28036 (use of our interrupts package).